repo2docker/buildpacks/conda/environment.py-3.6.yml

```diff
@@ -1,10 +1,10 @@
 # AUTO GENERATED FROM environment.yml, DO NOT MANUALLY MODIFY
-# Generated on 2019-09-13 19:47:35 UTC
+# Generated on 2020-01-29 16:07:07 UTC
 dependencies:
 - python=3.6.*
-- ipywidgets==7.4.2
-- jupyterlab==0.35.4
-- jupyterhub=1.0
-- nbconvert==5.4.1
-- notebook==5.7.8
+- ipywidgets==7.5.1
+- jupyterlab==1.2.6
+- jupyterhub-base==1.1.0
```

Comment:
Is this actually whats needed for any singleuser image when interacting with jupyterhub, as compared to the full jupyterhub? Wait what is jupyterhub-base? I didn't find it on PyPI

Reply:
The `jupyterhub` conda package depends on configurable-http-proxy (and thereby nodejs) because that's what is required to run jupyterhub. `jupyterhub-base` is a *conda* package that only contains the jupyterhub Python package, and not the full dependencies of jupyterhub. Basically, this is the same thing you would get if you did `pip install jupyterhub` since pip doesn't know about non-Python dependencies. A similar pattern is followed in e.g. `matplotlib-base` conda package, which has only the matplotlib Python package, but not dependencies required for 'default' use such as Qt.

Reply:
Ah makes sense, thank you! I've come to learn a lot about conda through you Min =)

Reply:
Yeah, it can be confusing when Python packages don't map 1:1 onto conda packages. You see the same thing with linux packages, e.g. what you might think of as just `libfoo` is split into `libfoo-dev` `libfoo-static`, `libfoo`, etc. Conda can do similar things, but doesn't always. It only does it when the separate installation burden is deemed "worth it" e.g. whether to pull in large additional dependencies like nodejs. The nice thing about the conda `juptyerhub` package is it 'just works' by default, unlike the Python one, because it comes with a proxy implementation. The downside is that the same package is used to provide single-user functionality, but doesn't need the proxy. Hence the split package

Reply:
Thanks for sharing this experience Min!